File: src/services/git-service.ts
/**
 * Git Service for TypeScript
 * Platform-independent Git operations
 */
import { spawn, spawnSync, SpawnSyncOptions } from 'child_process';
import * as path from 'path';
import * as fs from 'fs';
import * as os from 'os';

export interface GitStatus {
  branch: string;
  modified: string[];
  untracked: string[];
}

export interface GitResult {
  success: boolean;
  output?: string;
  error?: string;
}

export class GitService {
  private repoPath: string;
  private isWindows: boolean;

  constructor(repoPath: string = process.cwd()) {
    this.repoPath = path.resolve(repoPath);
    this.isWindows = os.platform() === 'win32';
    this.validateGitRepo();
  }

  private validateGitRepo(): void {
    const gitDir = path.join(this.repoPath, '.git');
    if (!fs.existsSync(gitDir)) {
      throw new Error(`${this.repoPath} is not a valid git repository`);
    }
  }

  private execute(args: string[]): { stdout: string; stderr: string; status: number } {
    const options: SpawnSyncOptions = {
      cwd: this.repoPath,
      encoding: 'utf8',
      // Windows에서 shell 사용
      shell: this.isWindows,
      // PATH 환경변수 명시적 설정
      env: {
        ...process.env,
        PATH: this.isWindows 
          ? `C:\\Windows\\System32;${process.env.PATH}` 
          : process.env.PATH
      }
    };

    try {
      const result = spawnSync('git', args, options);
      
      return {
        stdout: result.stdout?.toString() || '',
        stderr: result.stderr?.toString() || '',
        status: result.status || -1
      };
    } catch (error: any) {
      throw new Error(`Git command failed: git ${args.join(' ')}\nError: ${error.message}`);
    }
  }

  async status(): Promise<GitStatus> {
    const result = this.execute(['status', '-s']);
    
    if (result.status !== 0) {
      throw new Error(`git status failed: ${result.stderr}`);
    }

    const lines = result.stdout.split('\n').filter(line => line.trim());
    const modified: string[] = [];
    const untracked: string[] = [];

    lines.forEach(line => {
      if (line.startsWith(' M')) {
        modified.push(line.substring(3));
      } else if (line.startsWith('??')) {
        untracked.push(line.substring(3));
      }
    });

    const branch = this.getCurrentBranch();

    return { branch, modified, untracked };
  }

  getCurrentBranch(): string {
    const result = this.execute(['branch', '--show-current']);
    return result.status === 0 ? result.stdout.trim() : 'unknown';
  }

  async push(remote: string = 'origin', branch?: string): Promise<GitResult> {
    if (!branch) {
      branch = this.getCurrentBranch();
    }

    const result = this.execute(['push', remote, branch]);
    
    return {
      success: result.status === 0,
      output: result.stdout + result.stderr,
      error: result.status !== 0 ? result.stderr : undefined
    };
  }

  async commit(message: string): Promise<GitResult> {
    // 인코딩 문제 방지를 위한 임시 파일 사용
    const tmpFile = path.join(os.tmpdir(), `git-commit-${Date.now()}.txt`);
    
    try {
      fs.writeFileSync(tmpFile, message, 'utf8');
      const result = this.execute(['commit', '-F', tmpFile]);
      
      return {
        success: result.status === 0,
        output: result.stdout,
        error: result.status !== 0 ? result.stderr : undefined
      };
    } finally {
      // 임시 파일 삭제
      if (fs.existsSync(tmpFile)) {
        fs.unlinkSync(tmpFile);
      }
    }
  }

  async add(files?: string[]): Promise<GitResult> {
    const args = ['add'];
    if (files && files.length > 0) {
      args.push(...files);
    } else {
      args.push('-A');
    }

    const result = this.execute(args);
    
    return {
      success: result.status === 0,
      output: result.stdout,
      error: result.status !== 0 ? result.stderr : undefined
    };
  }
}

// Singleton instance
let gitServiceInstance: GitService | null = null;

export function getGitService(repoPath?: string): GitService {
  if (!gitServiceInstance) {
    gitServiceInstance = new GitService(repoPath);
  }
  return gitServiceInstance;
}

File: python/core/git_service.py
"""
Git Service Module
플랫폼 독립적인 Git 명령어 실행을 위한 서비스 모듈
"""
import subprocess
import os
import platform
from typing import Dict, List, Optional, Tuple
from pathlib import Path
import tempfile
import json


class GitService:
    """Git 명령어 실행을 추상화한 서비스 클래스"""
    
    def __init__(self, repo_path: str = "."):
        self.repo_path = Path(repo_path).resolve()
        self.is_windows = platform.system() == "Windows"
        self._validate_git_repo()
    
    def _validate_git_repo(self) -> None:
        """Git 저장소 유효성 검증"""
        git_dir = self.repo_path / ".git"
        if not git_dir.exists():
            raise ValueError(f"{self.repo_path} is not a valid git repository")
    
    def _execute(self, cmd: List[str], **kwargs) -> subprocess.CompletedProcess:
        """
        Git 명령어 실행 (플랫폼 독립적)
        
        Args:
            cmd: 실행할 명령어 리스트 (예: ['git', 'status'])
            **kwargs: subprocess.run에 전달할 추가 옵션
        
        Returns:
            subprocess.CompletedProcess 객체
        """
        # 기본 옵션 설정
        default_kwargs = {
            'capture_output': True,
            'text': True,
            'cwd': str(self.repo_path),
            'encoding': 'utf-8',
            'errors': 'ignore'
        }
        
        # Windows에서는 shell=True 사용
        if self.is_windows:
            default_kwargs['shell'] = True
        
        # 사용자 옵션으로 덮어쓰기
        default_kwargs.update(kwargs)
        
        try:
            result = subprocess.run(cmd, **default_kwargs)
            return result
        except Exception as e:
            # 상세한 오류 정보 포함
            raise RuntimeError(f"Git command failed: {' '.join(cmd)}\nError: {str(e)}")
    
    def status(self, short: bool = False) -> Dict[str, any]:
        """Git 상태 확인"""
        cmd = ['git', 'status']
        if short:
            cmd.append('-s')
        
        result = self._execute(cmd)
        
        if result.returncode != 0:
            raise RuntimeError(f"git status failed: {result.stderr}")
        
        # 결과 파싱
        if short:
            lines = result.stdout.strip().split('\n') if result.stdout else []
            modified = []
            untracked = []
            
            for line in lines:
                if line.startswith(' M'):
                    modified.append(line[3:])
                elif line.startswith('??'):
                    untracked.append(line[3:])
            
            return {
                'modified': modified,
                'untracked': untracked,
                'branch': self.get_current_branch()
            }
        
        return {'output': result.stdout}
    
    def get_current_branch(self) -> str:
        """현재 브랜치 이름 가져오기"""
        result = self._execute(['git', 'branch', '--show-current'])
        return result.stdout.strip() if result.returncode == 0 else 'unknown'
    
    def add(self, files: Optional[List[str]] = None) -> bool:
        """파일 스테이징"""
        cmd = ['git', 'add']
        
        if files:
            cmd.extend(files)
        else:
            cmd.append('-A')  # 모든 파일
        
        result = self._execute(cmd)
        return result.returncode == 0
    
    def commit(self, message: str) -> bool:
        """
        커밋 실행 (인코딩 안전)
        
        Args:
            message: 커밋 메시지
        
        Returns:
            성공 여부
        """
        # 인코딩 문제 방지를 위해 임시 파일 사용
        with tempfile.NamedTemporaryFile(mode='w', encoding='utf-8', 
                                        suffix='.txt', delete=False) as f:
            f.write(message)
            temp_file = f.name
        
        try:
            result = self._execute(['git', 'commit', '-F', temp_file])
            return result.returncode == 0
        finally:
            # 임시 파일 삭제
            if os.path.exists(temp_file):
                os.unlink(temp_file)
    
    def push(self, remote: str = 'origin', branch: Optional[str] = None) -> Tuple[bool, str]:
        """
        원격 저장소로 푸시
        
        Args:
            remote: 원격 저장소 이름
            branch: 브랜치 이름 (None이면 현재 브랜치)
        
        Returns:
            (성공여부, 출력메시지) 튜플
        """
        if branch is None:
            branch = self.get_current_branch()
        
        result = self._execute(['git', 'push', remote, branch])
        
        # Git은 종종 정보를 stderr로 출력
        output = result.stdout + result.stderr
        
        return (result.returncode == 0, output)
    
    def fetch(self, remote: str = 'origin') -> bool:
        """원격 저장소에서 가져오기"""
        result = self._execute(['git', 'fetch', remote])
        return result.returncode == 0
    
    def log(self, n: int = 5, oneline: bool = True) -> List[str]:
        """커밋 로그 조회"""
        cmd = ['git', 'log']
        
        if oneline:
            cmd.append('--oneline')
        
        cmd.extend(['-n', str(n)])
        
        result = self._execute(cmd)
        
        if result.returncode != 0:
            return []
        
        return result.stdout.strip().split('\n') if result.stdout else []
    
    def get_remote_url(self, remote: str = 'origin') -> Optional[str]:
        """원격 저장소 URL 가져오기"""
        result = self._execute(['git', 'remote', 'get-url', remote])
        return result.stdout.strip() if result.returncode == 0 else None


# 싱글톤 인스턴스
_git_service = None

def get_git_service(repo_path: str = ".") -> GitService:
    """GitService 싱글톤 인스턴스 반환"""
    global _git_service
    if _git_service is None:
        _git_service = GitService(repo_path)
    return _git_service
